ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/KeyedWindowedTableLookupOperator.java

```diff
@@ -0,0 +1,137 @@
+import io.confluent.ksql.physical.pull.operators.WhereInfo.WindowBounds;
+import io.confluent.ksql.planner.plan.DataSourceNode;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.kafka.connect.data.Struct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KeyedWindowedTableLookupOperator
+    extends AbstractPhysicalOperator
+    implements UnaryPhysicalOperator, DataSourceOperator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      KeyedWindowedTableLookupOperator.class);
+
+  private final Materialization mat;
+  private final DataSourceNode logicalOperator;
+  private final WindowBounds windowBounds;
+
+  private List<KsqlPartitionLocation> partitionLocations;
+  private Iterator<WindowedRow> resultIterator;
+  private Iterator<Struct> keyIterator;
+  private Iterator<KsqlPartitionLocation> partitionLocationIterator;
+  private KsqlPartitionLocation nextLocation;
+  private Struct nextKey;
+
+
+  public KeyedWindowedTableLookupOperator(
+      final Materialization mat,
+      final DataSourceNode logicalNode,
+      final WindowBounds windowBounds
+  ) {
+    this.logicalOperator = Objects.requireNonNull(logicalNode, "logicalNode");
+    this.mat = Objects.requireNonNull(mat, "mat");
+    this.windowBounds = Objects.requireNonNull(windowBounds, "windowBounds");
+  }
+
+  @Override
+  public void open() {
+    partitionLocationIterator = partitionLocations.iterator();
+    if (partitionLocationIterator.hasNext()) {
+      nextLocation = partitionLocationIterator.next();
+      if (!nextLocation.getKeys().isPresent()) {
+        throw new IllegalStateException("Table windowed queries should be done with keys");
+      }
+      keyIterator = nextLocation.getKeys().get().iterator();
+      if (keyIterator.hasNext()) {
+        nextKey = keyIterator.next();
+        resultIterator = mat.windowed().get(
+            nextKey,
+            nextLocation.getPartition(),
+            windowBounds.getStart(),
+            windowBounds.getEnd())
+            .iterator();
+      }
+    }
+  }
+
+  @Override
+  public Object next() {
+    if (resultIterator.hasNext()) {
+      return resultIterator.next();
+    }
+    // Exhausted resultIterator
+    if (!keyIterator.hasNext()) {
+      if (partitionLocationIterator.hasNext()) {
+        nextLocation = partitionLocationIterator.next();
+      } else {
+        // Exhausted all iterators
+        return null;
+      }
+      if (!nextLocation.getKeys().isPresent()) {
+        throw new IllegalStateException("Table lookup queries should be done with keys");
+      }
+      keyIterator = nextLocation.getKeys().get().iterator();
+    }
+    nextKey = keyIterator.next();
+    resultIterator = mat.windowed().get(
+        nextKey,
+        nextLocation.getPartition(),
+        windowBounds.getStart(),
+        windowBounds.getEnd())
+        .iterator();
+    if (resultIterator.hasNext()) {
```

Comment:
What if `resultIterator.hasNext()` is false, but `keyIterator.hasNext()` is true?  You would return null below when there might be more.  It seems like you could recursively call `next()` after setting up the `nextKey` and `resultIterator`.

Reply:
`resultIterator.hasNext()` is false, but `keyIterator.hasNext()` is true it will go to line 91 and get the next keyIterator `keyIterator.next()`. Am I missing something?

Reply:
If you were to call `this.next()` again, it would do what you're saying and see that there are more keys, but from what I can see, it doesn't do that by default (and hence my recursive comment).

Specifically, imagine you're looking up two keys, `1` and `2`.  If you do `resultIterator = mat.windowed().get(` .. for `1` and get nothing, `resultIterator.hasNext()` will be false, and you'll fall through to `return null` below. 

I guess maybe that's the correct if you want to return nulls for empty lookups, but then you have to do a null check with the caller.  I was effectively imagining that you would just continue on to the next non empty iterator.